pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/servlet/SecurityServletFilter.java

```diff
@@ -52,69 +62,90 @@
-            } else if (user.trim().isEmpty()) {
-                throw new IllegalArgumentException(EMPTY_HEADER_ERROR);
+            final String gpdbUser = getHeaderValue(request, USER_HEADER);
+            String transactionId = getHeaderValue(request, TRANSACTION_ID_HEADER);
+            Integer segmentId = getHeaderValueInt(request, SEGMENT_ID_HEADER, true);
+            Integer fragmentCount = getHeaderValueInt(request, FRAGMENT_COUNT_HEADER, false);
+            Integer fragmentIndex = getHeaderValueInt(request, FRAGMENT_INDEX_HEADER, false);
+
+            SessionId session = new SessionId(segmentId, transactionId, gpdbUser);
+            if (LOG.isDebugEnabled() && fragmentCount != null) {
+                LOG.debug(session.toString() + " Fragment = " + fragmentIndex + " of " + fragmentCount);
             }
 
             // TODO refresh Kerberos token when security is enabled
 
-            // prepare pivileged action to run on behalf of proxy user
+            // prepare privileged action to run on behalf of proxy user
             PrivilegedExceptionAction<Boolean> action = new PrivilegedExceptionAction<Boolean>() {
                 @Override
                 public Boolean run() throws IOException, ServletException {
-                    LOG.debug("Performing request chain call for proxy user = " + user);
+                    LOG.debug("Performing request chain call for proxy user = " + gpdbUser);
                     chain.doFilter(request, response);
                     return true;
                 }
             };
 
             // create proxy user UGI from the UGI of the logged in user and execute the servlet chain as that user
-            UserGroupInformation proxyUGI = null;
+            UserGroupInformation ugi = cache.getUserGroupInformation(session);
             try {
-                LOG.debug("Creating proxy user = " + user);
-                proxyUGI = UserGroupInformation.createProxyUser(user, UserGroupInformation.getLoginUser());
-                proxyUGI.doAs(action);
+                ugi.doAs(action);
             } catch (UndeclaredThrowableException ute) {
                 // unwrap the real exception thrown by the action
                 throw new ServletException(ute.getCause());
             } catch (InterruptedException ie) {
                 throw new ServletException(ie);
             } finally {
-                try {
-                    if (proxyUGI != null) {
-                        LOG.debug("Closing FileSystem for proxy user = " + proxyUGI.getUserName());
-                        FileSystem.closeAllForUGI(proxyUGI);
-                    }
-                } catch (Throwable t) {
-                    LOG.warn("Error closing FileSystem for proxy user = " + proxyUGI.getUserName());
-                }
+                // Optimization to cleanup the cache if it is the last fragment
+                boolean forceClean = (fragmentIndex != null && fragmentIndex.equals(fragmentCount));
```

Comment:
Why would fragmentIndex matter here ? Are we certain that the requests are always receieved in the order of fragment index ?

Reply:
For the same segment, they are always called in order.